// @azure-tools/typespec-autorest/dist/src/decorators.js
import { createStateSymbol, reportDiagnostic } from "./lib.js";
export const namespace = "Autorest";
const exampleKey = createStateSymbol("example");
/**
 * `@example` - attaches example files to an operation. Multiple examples can be specified.
 *
 * @param {string} param pathOrUri - path or Uri to the example file.
 * @param {string} param title - name or description of the example file.
 *
 * `@example` can be specified on Operations.
 */
export function $example(context, entity, pathOrUri, title) {
    const { program } = context;
    if (!program.stateMap(exampleKey).has(entity)) {
        program.stateMap(exampleKey).set(entity, []);
    }
    else if (program
        .stateMap(exampleKey)
        .get(entity)
        .find((e) => e.title === title || e.pathOrUri === pathOrUri)) {
        reportDiagnostic(program, {
            code: "duplicate-example",
            target: entity,
        });
    }
    program.stateMap(exampleKey).get(entity).push({
        pathOrUri,
        title,
    });
}
export function getExamples(program, entity) {
    return program.stateMap(exampleKey).get(entity);
}
const refTargetsKey = createStateSymbol("autorest.ref");
/**
 * `@useRef` - is used to replace the TypeSpec model type in emitter output with a pre-existing named OpenAPI schema such as ARM common types.
 *
 * @param {string} param jsonRef - path or Uri to an OpenAPI schema.
 *
 * `@useRef` can be specified on Models and ModelProperty.
 */
export function $useRef(context, entity, jsonRef) {
    context.program.stateMap(refTargetsKey).set(entity, jsonRef);
}
export function getRef(program, entity) {
    const refOrProducer = program.stateMap(refTargetsKey).get(entity);
    return refOrProducer;
}
